// ng-zorro-antd/bundles/ng-zorro-antd-page-header.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('ng-zorro-antd/core'), require('ng-zorro-antd/icon')) :
    typeof define === 'function' && define.amd ? define('ng-zorro-antd/page-header', ['exports', '@angular/common', '@angular/core', 'ng-zorro-antd/core', 'ng-zorro-antd/icon'], factory) :
    (global = global || self, factory((global['ng-zorro-antd'] = global['ng-zorro-antd'] || {}, global['ng-zorro-antd']['page-header'] = {}), global.ng.common, global.ng.core, global['ng-zorro-antd'].core, global['ng-zorro-antd'].icon));
}(this, function (exports, common, core, core$1, icon) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NzPageHeaderTitleDirective = /** @class */ (function () {
        function NzPageHeaderTitleDirective() {
        }
        NzPageHeaderTitleDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-page-header-title, [nz-page-header-title]',
                        exportAs: 'nzPageHeaderTitle',
                        host: {
                            class: 'ant-page-header-heading-title'
                        }
                    },] }
        ];
        return NzPageHeaderTitleDirective;
    }());
    var NzPageHeaderSubtitleDirective = /** @class */ (function () {
        function NzPageHeaderSubtitleDirective() {
        }
        NzPageHeaderSubtitleDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-page-header-subtitle, [nz-page-header-subtitle]',
                        exportAs: 'nzPageHeaderSubtitle',
                        host: {
                            class: 'ant-page-header-heading-sub-title'
                        }
                    },] }
        ];
        return NzPageHeaderSubtitleDirective;
    }());
    var NzPageHeaderContentDirective = /** @class */ (function () {
        function NzPageHeaderContentDirective() {
        }
        NzPageHeaderContentDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-page-header-content, [nz-page-header-content]',
                        exportAs: 'nzPageHeaderContent',
                        host: {
                            class: 'ant-page-header-content'
                        }
                    },] }
        ];
        return NzPageHeaderContentDirective;
    }());
    var NzPageHeaderTagDirective = /** @class */ (function () {
        function NzPageHeaderTagDirective() {
        }
        NzPageHeaderTagDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-page-header-tags, [nz-page-header-tags]',
                        exportAs: 'nzPageHeaderTags',
                        host: {
                            class: 'ant-page-header-heading-tags'
                        }
                    },] }
        ];
        return NzPageHeaderTagDirective;
    }());
    var NzPageHeaderExtraDirective = /** @class */ (function () {
        function NzPageHeaderExtraDirective() {
        }
        NzPageHeaderExtraDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-page-header-extra, [nz-page-header-extra]',
                        exportAs: 'nzPageHeaderExtra',
                        host: {
                            class: 'ant-page-header-heading-extra'
                        }
                    },] }
        ];
        return NzPageHeaderExtraDirective;
    }());
    var NzPageHeaderFooterDirective = /** @class */ (function () {
        function NzPageHeaderFooterDirective() {
        }
        NzPageHeaderFooterDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-page-header-footer, [nz-page-header-footer]',
                        exportAs: 'nzPageHeaderFooter',
                        host: {
                            class: 'ant-page-header-footer'
                        }
                    },] }
        ];
        return NzPageHeaderFooterDirective;
    }());
    var NzPageHeaderBreadcrumbDirective = /** @class */ (function () {
        function NzPageHeaderBreadcrumbDirective() {
        }
        NzPageHeaderBreadcrumbDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-breadcrumb[nz-page-header-breadcrumb]',
                        exportAs: 'nzPageHeaderBreadcrumb'
                    },] }
        ];
        return NzPageHeaderBreadcrumbDirective;
    }());
    var NzPageHeaderAvatarDirective = /** @class */ (function () {
        function NzPageHeaderAvatarDirective() {
        }
        NzPageHeaderAvatarDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'nz-avatar[nz-page-header-avatar]',
                        exportAs: 'nzPageHeaderAvatar'
                    },] }
        ];
        return NzPageHeaderAvatarDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var NzPageHeaderComponent = /** @class */ (function () {
        function NzPageHeaderComponent(location) {
            this.location = location;
            this.isTemplateRefBackIcon = false;
            this.isStringBackIcon = false;
            this.nzBackIcon = null;
            this.nzBack = new core.EventEmitter();
        }
        /**
         * @param {?} changes
         * @return {?}
         */
        NzPageHeaderComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
        function (changes) {
            if (changes.hasOwnProperty('nzBackIcon')) {
                this.isTemplateRefBackIcon = changes.nzBackIcon.currentValue instanceof core.TemplateRef;
                this.isStringBackIcon = typeof changes.nzBackIcon.currentValue === 'string';
            }
        };
        /**
         * @return {?}
         */
        NzPageHeaderComponent.prototype.onBack = /**
         * @return {?}
         */
        function () {
            if (this.nzBack.observers.length) {
                this.nzBack.emit();
            }
            else {
                this.location.back();
            }
        };
        NzPageHeaderComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'nz-page-header',
                        exportAs: 'nzPageHeader',
                        template: "<ng-content select=\"nz-breadcrumb[nz-page-header-breadcrumb]\"></ng-content>\n\n<div class=\"ant-page-header-heading\">\n  <!--back-->\n  <div *ngIf=\"nzBackIcon !== null\" (click)=\"onBack()\" class=\"ant-page-header-back\">\n    <div role=\"button\" tabindex=\"0\" class=\"ant-page-header-back-button\">\n      <i *ngIf=\"isStringBackIcon\" nz-icon [nzType]=\"nzBackIcon ? nzBackIcon : 'arrow-left'\" nzTheme=\"outline\"></i>\n      <ng-container *ngIf=\"isTemplateRefBackIcon\" [ngTemplateOutlet]=\"nzBackIcon\"></ng-container>\n    </div>\n  </div>\n  <!--avatar-->\n  <ng-content select=\"nz-avatar[nz-page-header-avatar]\"></ng-content>\n  <!--title-->\n  <span class=\"ant-page-header-heading-title\" *ngIf=\"nzTitle\">\n    <ng-container *nzStringTemplateOutlet=\"nzTitle\">{{ nzTitle }}</ng-container>\n  </span>\n  <ng-content *ngIf=\"!nzTitle\" select=\"nz-page-header-title, [nz-page-header-title]\"></ng-content>\n  <!--subtitle-->\n  <span class=\"ant-page-header-heading-sub-title\" *ngIf=\"nzSubtitle\">\n    <ng-container *nzStringTemplateOutlet=\"nzSubtitle\">{{ nzSubtitle }}</ng-container>\n  </span>\n  <ng-content *ngIf=\"!nzSubtitle\" select=\"nz-page-header-subtitle, [nz-page-header-subtitle]\"></ng-content>\n  <ng-content select=\"nz-page-header-tags, [nz-page-header-tags]\"></ng-content>\n  <ng-content select=\"nz-page-header-extra, [nz-page-header-extra]\"></ng-content>\n</div>\n\n<ng-content select=\"nz-page-header-content, [nz-page-header-content]\"></ng-content>\n<ng-content select=\"nz-page-header-footer, [nz-page-header-footer]\"></ng-content>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        encapsulation: core.ViewEncapsulation.None,
                        host: {
                            class: 'ant-page-header',
                            '[class.ant-page-header-has-footer]': 'nzPageHeaderFooter'
                        },
                        styles: ["nz-page-header,nz-page-header-content,nz-page-header-footer{display:block}", "\n      .ant-page-header-back-button {\n        border: 0px;\n        background: transparent;\n        padding: 0px;\n        line-height: inherit;\n        display: inline-block;\n      }\n    "]
                    }] }
        ];
        /** @nocollapse */
        NzPageHeaderComponent.ctorParameters = function () { return [
            { type: common.Location }
        ]; };
        NzPageHeaderComponent.propDecorators = {
            nzBackIcon: [{ type: core.Input }],
            nzTitle: [{ type: core.Input }],
            nzSubtitle: [{ type: core.Input }],
            nzBack: [{ type: core.Output }],
            nzPageHeaderFooter: [{ type: core.ContentChild, args: [NzPageHeaderFooterDirective, { static: false },] }]
        };
        return NzPageHeaderComponent;
    }());
    if (false) {
        /** @type {?} */
        NzPageHeaderComponent.prototype.isTemplateRefBackIcon;
        /** @type {?} */
        NzPageHeaderComponent.prototype.isStringBackIcon;
        /** @type {?} */
        NzPageHeaderComponent.prototype.nzBackIcon;
        /** @type {?} */
        NzPageHeaderComponent.prototype.nzTitle;
        /** @type {?} */
        NzPageHeaderComponent.prototype.nzSubtitle;
        /** @type {?} */
        NzPageHeaderComponent.prototype.nzBack;
        /** @type {?} */
        NzPageHeaderComponent.prototype.nzPageHeaderFooter;
        /**
         * @type {?}
         * @private
         */
        NzPageHeaderComponent.prototype.location;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var NzPageHeaderCells = [
        NzPageHeaderTitleDirective,
        NzPageHeaderSubtitleDirective,
        NzPageHeaderContentDirective,
        NzPageHeaderTagDirective,
        NzPageHeaderExtraDirective,
        NzPageHeaderFooterDirective,
        NzPageHeaderBreadcrumbDirective,
        NzPageHeaderAvatarDirective
    ];
    var NzPageHeaderModule = /** @class */ (function () {
        function NzPageHeaderModule() {
        }
        NzPageHeaderModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule, core$1.NzAddOnModule, icon.NzIconModule],
                        exports: [NzPageHeaderComponent, NzPageHeaderCells],
                        declarations: [NzPageHeaderComponent, NzPageHeaderCells]
                    },] }
        ];
        return NzPageHeaderModule;
    }());

    exports.NzPageHeaderAvatarDirective = NzPageHeaderAvatarDirective;
    exports.NzPageHeaderBreadcrumbDirective = NzPageHeaderBreadcrumbDirective;
    exports.NzPageHeaderComponent = NzPageHeaderComponent;
    exports.NzPageHeaderContentDirective = NzPageHeaderContentDirective;
    exports.NzPageHeaderExtraDirective = NzPageHeaderExtraDirective;
    exports.NzPageHeaderFooterDirective = NzPageHeaderFooterDirective;
    exports.NzPageHeaderModule = NzPageHeaderModule;
    exports.NzPageHeaderSubtitleDirective = NzPageHeaderSubtitleDirective;
    exports.NzPageHeaderTagDirective = NzPageHeaderTagDirective;
    exports.NzPageHeaderTitleDirective = NzPageHeaderTitleDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
